File: CryptoBridge-python/src/presentation/api/routers/bridge_router.py
from fastapi import APIRouter, Depends, HTTPException
from typing import Union
from ...schemas.bridge_schemas import (
    TickerRequest, SymbolRequest, TickersResponse,
    SymbolsResponse, ExchangesResponse, ErrorResponse
)
from ..dependencies import get_bridge_service
from ....application.services.bridge_service import BridgeService

router = APIRouter()


@router.get("/", response_model=ExchangesResponse)
async def get_exchanges(
    bridge_service: BridgeService = Depends(get_bridge_service)
):
    try:
        result = await bridge_service.get_exchanges()
        return result
    except Exception as e:
        print(f"Erro detalhado ao buscar exchanges: {str(e)}")
        return {
            "status": 500,
            "msg": "Erro interno do servidor."
        }


@router.post("/symbols", response_model=Union[SymbolsResponse, ErrorResponse])
async def get_symbols(
    request: SymbolRequest,
    bridge_service: BridgeService = Depends(get_bridge_service)
):
    try:
        result = await bridge_service.get_symbols(request.exchange)
        return result
    except Exception as e:
        print(f"Erro detalhado ao buscar símbolos: {str(e)}")
        return {
            "status": 500,
            "msg": "Erro interno do servidor."
        }


@router.post("/tickers", response_model=Union[TickersResponse, ErrorResponse])
async def get_tickers(
    request: TickerRequest,
    bridge_service: BridgeService = Depends(get_bridge_service)
):
    try:
        result = await bridge_service.get_tickers(request.exchange, request.tickers)
        return result
    except Exception as e:
        print(f"Erro detalhado ao buscar tickers: {str(e)}")
        return {
            "status": 500,
            "msg": "Erro interno do servidor."
        }

File: CryptoBridge-node/src/presentation/routes/BridgeRouter.js
const express = require('express');
const Joi = require('joi');

const router = express.Router();

// Validation schemas
const symbolsSchema = Joi.object({
    exchange: Joi.string().min(1).required()
});

const tickersSchema = Joi.object({
    exchange: Joi.string().min(1).required(),
    tickers: Joi.array().items(Joi.string().min(1)).min(1).required()
});

class BridgeRouter {
    constructor(bridgeService) {
        this.bridgeService = bridgeService;
        this.setupRoutes();
    }

    setupRoutes() {
        // GET / - Lista exchanges disponíveis
        router.get('/', async (req, res) => {
            try {
                const result = await this.bridgeService.getExchanges();
                res.json(result);
            } catch (error) {
                console.error('Erro detalhado ao buscar exchanges:', error);
                res.status(500).json({
                    status: 500,
                    msg: 'Erro interno do servidor.'
                });
            }
        });

        // POST /symbols - Lista símbolos de uma exchange
        router.post('/symbols', async (req, res) => {
            try {
                // Validate request body
                const { error, value } = symbolsSchema.validate(req.body);
                if (error) {
                    return res.status(422).json({
                        status: 422,
                        msg: 'Dados de entrada inválidos.',
                        details: error.details
                    });
                }

                const result = await this.bridgeService.getSymbols(value.exchange);
                res.json(result);
            } catch (error) {
                console.error('Erro detalhado ao buscar símbolos:', error);
                res.status(500).json({
                    status: 500,
                    msg: 'Erro interno do servidor.'
                });
            }
        });

        // POST /tickers - Busca preços de tickers
        router.post('/tickers', async (req, res) => {
            try {
                // Validate request body
                const { error, value } = tickersSchema.validate(req.body);
                if (error) {
                    return res.status(422).json({
                        status: 422,
                        msg: 'Dados de entrada inválidos.',
                        details: error.details
                    });
                }

                const result = await this.bridgeService.getTickers(value.exchange, value.tickers);
                res.json(result);
            } catch (error) {
                console.error('Erro detalhado ao buscar tickers:', error);
                res.status(500).json({
                    status: 500,
                    msg: 'Erro interno do servidor.'
                });
            }
        });
    }

    getRouter() {
        return router;
    }
}

module.exports = BridgeRouter;